### PR TITLE
Add GroupVersion tags to OpenAPI spec and remove all specs except main one

### PR DESCRIPTION
**Pull request by mbohlool**
Tags are used as a grouping mechanism in OpenAPI. We generated one spec per GroupVersion before for this grouping but by adding those tags in this PR, those files have no use. We can always add them back if there were a use-case for them.

**Release note**:

``` release-note
Deprecate OpenAPI spec for GroupVersion endpoints in favor of single spec /swagger.json
```

Reference: #13414

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35388)

### CODE REVIEW

**Comment by mbohlool** [comment]
cc @bgrant0607 @thockin @lavalamp @kubernetes/sig-api-machinery  

Please look at this change as it is a design change.

File: hack/update-federation-openapi-spec.sh [hunk 1]
```diff
@@ -60,18 +60,15 @@ kube::log::status "Starting federation-apiserver"
   --insecure-port="${API_PORT}" \
   --etcd-servers="http://${ETCD_HOST}:${ETCD_PORT}" \
   --advertise-address="10.10.10.10" \
+  --cert-dir="${TMP_DIR}/certs" \
```
Comment (mbohlool) [inline hack/update-federation-openapi-spec.sh]: @nikhiljindal this is the reason federation server did not start on my machine.

File: pkg/apiserver/openapi/openapi.go [hunk 2]
```diff
@@ -72,10 +73,15 @@ func GetOperationID(servePath string, r *restful.Route) (string, error) {
 			prefix = prefix + ToValidOperationID(strings.TrimSuffix(parts[1], ".k8s.io"), prefix != "")
 			if len(parts) > 2 {
 				prefix = prefix + ToValidOperationID(parts[2], prefix != "")
+				tags = append(tags, fmt.Sprintf("%s_%s", ToValidOperationID(parts[2], false), ToValidOperationID(strings.TrimSuffix(parts[1], ".k8s.io"), false)))
```
Comment (caesarxuchao) [inline pkg/apiserver/openapi/openapi.go]: Why version_group rather than group_version?

**Review by caesarxuchao — COMMENTED** [review]
`+2,012 −48,660` is cool! Just curious, I thought we just merged the swagger into one file, why would the PR remove so many lines? Is it because there were many duplicate swagger definition before?

Also out of curiosity, when generating code, will the swagger paths with the same tags be generated as one file? What generating tool are we using?

**Comment by smarterclayton** [comment]
I *think* we should always do GroupVersion because Group dominates Version.

File: hack/update-federation-openapi-spec.sh (hunk 1 above)
Comment (nikhiljindal) [inline hack/update-federation-openapi-spec.sh]: ack, thx!

File: pkg/apiserver/openapi/openapi.go (hunk 2 above)
Comment (nikhiljindal) [inline pkg/apiserver/openapi/openapi.go]: I dont see a reason for version_group. I think `group_version` or `group/version` or `group.version` should be fine which is what we use at other places.

**Comment by nikhiljindal** [comment]
LGTM once the group_version or version_group comment is resolved. Thx!

File: pkg/apiserver/openapi/openapi.go (hunk 2 above)
Comment (caesarxuchao) [inline pkg/apiserver/openapi/openapi.go]: Please don't use `.` as the separator. Group might have `.` in the name. (Though I remember you removed .k8s.io suffix)

Comment (mbohlool) [inline pkg/apiserver/openapi/openapi.go]: Yeah, I've used group_version.

**Comment by mbohlool** [comment]
@caesarxuchao We were removing swagger.json for all endpoints as well as a root one. This change will only keep root one and remove all those specs for endpoints, that why we have a lot of removed line.

Generators (at least the one I've tried) generate one file (or class) for each tag.

**Comment by caesarxuchao** [comment]
LGTM. Thanks.

**Comment by liggitt** [comment]
Does this mean we no longer serve openapi specs per group? I thought we were trying to allow decentralizing discovery and schema info

**Comment by mbohlool** [comment]
It is easy to convert the existing spec to any of the group spec (group them by tag). If we see a usecase for it, we can always bring it back (or write a simple tool to do the conversion) but we think it is easier to maintain this way. It is also possible (with little effort) to generate a client and pick the parts that you need (each tag will result in an API file. Before we had N spec and N API files, right now we have one spec and still N API files. If you only need core_v1, you can pick ApiCoreVi1.py (for python for example) the remove the rest.

**Comment by liggitt** [comment]
I'm more concerned about compatibility implications for clients that were fetching per group/version specs for validation (like kubectl). I couldn't tell if this was changing what was served by the API server.

**Comment by mbohlool** [comment]
Oh, that is still swagger 1.2 that  clients such as kubectl using and that is not changed. We are going to keep them for a while until we upgrade all clients to OpenAPI. 
This change is only about OpenAPI group specific specs, and they were added in the 1.5 development timeline, so they've never been release. Actually that was one of the reasons we removed them. We do not use them, we do not see any usecase for them, and it would be a compatibility issue to remove them after 1.5 release.

**Pull request merged**